[GSoC-2025] Blog post 4: Improved Console Output for Swift Testing (#1213)
apple/swift-org-website: _posts/2025-11-14-gsoc-2025-showcase-swift-testing-output.md

---
layout: new-layouts/post
published: true

> date: 2025-11-14 10:00:00

Reviewer: ```suggestion
date: 2025-11-14 17:15:00
```

title: "Swift GSoC 2025 Showcase: Improved Console Output for Swift Testing"
author: [kelvin, ktoso]
category: "Community"
---

The Swift community participated in [Google Summer of Code](https://summerofcode.withgoogle.com) 2025, and we've recently been showcasing all of the projects and work accomplished here on the Swift blog. You can learn more by following these convenient links:

- [Bringing Swiftly support to VS Code](/blog/gsoc-2025-showcase-swiftly-support-in-vscode/)
- [Extending Swift-Java Interoperability](/blog/gsoc-2025-showcase-swift-java/)
- [Improved code completion for Swift](/blog/gsoc-2025-showcase-code-completion/)
- Improved console output for Swift Testing (this post)

> Each GSoC contributor has shared a writeup about their project and experience in the program on the forums. The fourth project, and last in this year's series, improved the console output that is printed when running tests implemented with Swift Testing, contributed by Kelvin Bui.

Reviewer: ```suggestion
Each GSoC contributor has shared a writeup about their project and experience in the program. The fourth and last post in this year's series, contributed by Kelvin Bui, improved the console output that is printed when running tests implemented with Swift Testing.
```

### Problem & Motivation

The default console output for `swift test` was functional but presented several challenges for developers working on large and complex projects:

* **Difficulty Locating Failures:** In large test suites, critical failure messages were often lost in a long stream of success outputs, making it hard to quickly assess the health of a test run.
* **Lack of Structure:** The flat log format made it difficult to understand the relationship between tests and their parent suites, especially in projects with many nested modules.

Our vision was to solve these problems by creating an advanced console reporter that is both informative and intuitive.

> ## Key Technical Achievements

Reviewer: ```suggestion
### Key Technical Achievements
```

**New Hierarchical Summary:**
![New Hierarchical Test Results](/assets/images/gsoc-25/testing-output-2.jpg)

The new recorder introduces a rich, hierarchical display that immediately brings order and clarity to test output. Using Unicode box-drawing characters (with ASCII fallback), it clearly visualizes the nested structure of test modules, suites, and individual tests. This not only makes the test results significantly easier to read and navigate but also provides crucial context around failed tests, displaying detailed issue descriptions directly within their respective hierarchical nodes. This transformation dramatically improves the developer experience by offering instant insights into test organization and failure points.

![Failed Tests Showcase](/assets/images/gsoc-25/testing-output-3.jpg)

> ## Challenges & Lessons Learned

Reviewer: ```suggestion
### Challenges & Lessons Learned
```

* Deciding what to show inline vs. in a details section is a UX trade-off; too much inline detail makes the tree noisy, and too little hides useful context. The two-tier approach (short inline message + full details section) balances these needs.

**Collaboration & process lessons:**

* Presenting to the workgroup matters: Early live presentation to the Swift Testing workgroup exposed design assumptions quickly and gave direction that influenced major architectural choices.
* Iterative, small PRs win: Breaking work into digestible PRs (skeleton → ABI pivot → UI features) made reviews actionable and reduced reviewer fatigue.
* Incorporating targeted reviewer comments (e.g., ABI pivot) showed the importance of documenting design decisions (why we choose serialization, how the tree is constructed), which speeds later onboarding and review cycles.

**Personal growth:**

I learned how to accept and act on high-quality feedback from senior engineers, and how to reframe technical trade-offs as testable hypotheses. The project sharpened my ability to move from mockups to production code while keeping maintainability and cross-platform compatibility as first-class concerns.

> ## Acknowledgements

Reviewer: ```suggestion
### Acknowledgements
```
